### ISSUE TITLE
BOT driver: unexpected CSW received during SCP INQUIRY command.

### ISSUE THREAD

**Issue by duckbill007** [DarkMatterCore/libusbhsfs#4]
1. Plug-in USB drive.
2. Run hbl
3. Run example

Example hangs and do not react on any buttons.
If plug out flash drive - then receive exit signal and example exits.

Example works only if first run it and only after that insert usb drive.

**Comment by DarkMatterCore** [comment]
Works fine for my testers (I don't have a Switch atm).

Could you please build the example test application against the debug version of the library (`-lusbhsfs` -> `-lusbhsfsd`) and share its logfile?

**Comment by DarkMatterCore** [comment]
... Unless you're using an exFAT USB drive. For some reason, FatFs takes a long time to determine the free number of clusters under exFAT.

This has only been reported once, though, and the logfile only displayed lots of concurrent 1-sector reads generated by `f_getfree()`. I tried to see what's wrong, but I saw no issues within FatFs, which led me to believe it was some sort of filesystem corruption issue (exFAT has no backup FAT).

**Comment by duckbill007** [comment]
Yes. It is an exfat drive.

**Comment by DarkMatterCore** [comment]
Alright, then try commenting the `statvfs()` section from the example test application. It should confirm my suspicion.

**Comment by duckbill007** [comment]
Before commenting out statvfs()

Log for start example with USB drive inserted:
https://pastebin.com/SRB1Wyfe

Log for start with inserted/removing and reinsert:
https://pastebin.com/RYxKpMzJ

**Comment by duckbill007** [comment]
commenting `if (!statvfs(path, &fsinfo))
    {
        u64 total_size = ((u64)fsinfo.f_blocks * (u64)fsinfo.f_frsize);
        u64 free_space = ((u64)fsinfo.f_bfree * (u64)fsinfo.f_frsize);
        
        printf("OK!\n\t\t\t- Total FS size: 0x%lX bytes.\n\t\t\t- Free FS space: 0x%lX bytes.\n", total_size, free_space);
    } else {
        printf("FAILED! (%d).\n", errno);
    }` changed nothing

**Comment by DarkMatterCore** [comment]
The pastebin expired. Or so it seems.

**Comment by DarkMatterCore** [comment]
If you want to and/or think it'll be better, feel free to reach me via Discord. You can find my Discord handle in my GitHub profile.

**Comment by duckbill007** [comment]
First log:

________________________________________________________________
2020-11-24 17:44:55 -> usbHsFsInitialize: libusbhsfs v0.0.2 starting. Built on Nov 24 2020 - 10:53:47.
2020-11-24 17:44:55 -> usbHsFsDriveManagerThreadFunc: Interface available event triggered.
2020-11-24 17:44:55 -> usbHsFsUpdateDriveContexts: Current drive count: 0.
2020-11-24 17:44:55 -> usbHsFsUpdateDriveContexts: usbHsQueryAvailableInterfaces returned 1 interface(s) matching our filter.
2020-11-24 17:44:55 -> usbHsFsUpdateDriveContexts: Interface #0 (1342767104) data:
0000095013000000010000000904000003080650000000000000000000000000000000070582020002010705830340000800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000070501020002010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000487344657669636531392d2f54322f50312f41303100000000000000000000000000000000000000000000000000000000000000000000000000000000000000030000001300000012010002000000401c1bb10a0001010203010902270001010080310000000000835cc74ab00e0000
2020-11-24 17:44:55 -> usbHsFsAddDriveContextToList: Adding drive context for interface 1342767104.
2020-11-24 17:44:55 -> usbHsFsAddDriveContextToList: Reallocating drive context buffer from 0 to 1 (interface 1342767104).
2020-11-24 17:44:55 -> usbHsFsDriveInitializeContext: Max LUN count: 1 (interface 1342767104).
2020-11-24 17:44:55 -> usbHsFsScsiSendInquiryCommand: Sending command (interface 1342767104, LUN 0).
2020-11-24 17:44:55 -> usbHsFsScsiSendCommandBlockWrapper: Data from CBW to send (interface 1342767104, LUN 0):
555342439b6ecd062400000080000612000000240000000000000000000000
2020-11-24 17:44:55 -> usbHsFsScsiTransferCommand: usbHsFsRequestPostBuffer transferred 0xD byte(s), expected 0x24! (interface 1342767104, LUN 0).
2020-11-24 17:44:55 -> usbHsFsScsiStartDriveLogicalUnit: Inquiry failed! (interface 1342767104, LUN 0).
2020-11-24 17:44:55 -> usbHsFsDriveInitializeContext: Failed to initialize context for LUN #0! (interface 1342767104).
2020-11-24 17:44:55 -> usbHsFsDriveInitializeContext: Failed to initialize any LUN/filesystem contexts! (interface 1342767104).
2020-11-24 17:44:55 -> usbHsFsAddDriveContextToList: Freed drive context buffer.
2020-11-24 17:44:55 -> usbHsFsUpdateDriveContexts: Failed to add drive with ID 1342767104 to drive context list.
2020-11-24 17:44:55 -> usbHsFsUpdateDriveContexts: Added 0 drive context(s).
2020-11-24 17:44:55 -> usbHsFsDriveManagerThreadFunc: Interface state change event triggered.
2020-11-24 17:44:55 -> usbHsFsUpdateDriveContexts: Current drive count: 0.
2020-11-24 17:45:03 -> usbHsFsCloseDriveManagerThread: Signaling drive manager thread exit event...
2020-11-24 17:45:03 -> usbHsFsDriveManagerThreadFunc: Exit event triggered.
2020-11-24 17:45:03 -> usbHsFsCloseDriveManagerThread: Thread successfully closed.

**Comment by DarkMatterCore** [comment]
@duckbill007 Thanks. That helps a lot. Seems to be a BOT driver failure. Could you please help me test some changes to it?

**Comment by duckbill007** [comment]
Of course

**Comment by DarkMatterCore** [comment]
@duckbill007 Please try this debug build. It will generate additional information about that transfer mismatch during the SCSI INQUIRY command.

[libusbhsfs-example.zip](https://github.com/DarkMatterCore/libusbhsfs/files/5591323/libusbhsfs-example.zip)

**Comment by DarkMatterCore** [comment]
By the way, if you could also share the second log (where I suppose the hang was happening), I'd be grateful. I want to check that as well.

**Comment by duckbill007** [comment]
OK. I'll test it tomorrow morning.

**Comment by duckbill007** [comment]
Here is log from your attached example:

________________________________________________________________
2020-11-25 09:33:05 -> usbHsFsInitialize: libusbhsfs v0.0.2 starting. Built on Nov 24 2020 - 12:30:54.
2020-11-25 09:33:05 -> usbHsFsDriveManagerThreadFunc: Interface available event triggered.
2020-11-25 09:33:05 -> usbHsFsUpdateDriveContexts: Current drive count: 0.
2020-11-25 09:33:05 -> usbHsFsUpdateDriveContexts: usbHsQueryAvailableInterfaces returned 1 interface(s) matching our filter.
2020-11-25 09:33:05 -> usbHsFsUpdateDriveContexts: Interface #0 (1343225856) data:
000010506a00000001000000090400000308065000000000000000000000000000000007058202000201070583034000080000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007050102000201000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000048734465766963653130362d2f54322f50312f413031000000000000000000000000000000000000000000000000000000000000000000000000000000000000030000006a00000012010002000000401c1bb10a000101020301090227000101008031000000000096f24e5071420000
2020-11-25 09:33:05 -> usbHsFsAddDriveContextToList: Adding drive context for interface 1343225856.
2020-11-25 09:33:05 -> usbHsFsAddDriveContextToList: Reallocating drive context buffer from 0 to 1 (interface 1343225856).
2020-11-25 09:33:05 -> usbHsFsDriveInitializeContext: Max LUN count: 1 (interface 1343225856).
2020-11-25 09:33:05 -> usbHsFsScsiSendInquiryCommand: Sending command (interface 1343225856, LUN 0).
2020-11-25 09:33:05 -> usbHsFsScsiSendCommandBlockWrapper: Data from CBW to send (interface 1343225856, LUN 0):
55534243928b88342400000080000612000000240000000000000000000000
2020-11-25 09:33:05 -> usbHsFsScsiTransferCommand: usbHsFsRequestPostBuffer transferred 0xD byte(s), expected 0x24! (interface 1343225856, LUN 0).
2020-11-25 09:33:05 -> usbHsFsScsiTransferCommand: Data from unexpected CSW (interface 1343225856, LUN 0):
55534253928b88340000000000
2020-11-25 09:33:05 -> usbHsFsScsiStartDriveLogicalUnit: Inquiry failed! (interface 1343225856, LUN 0).
2020-11-25 09:33:05 -> usbHsFsDriveInitializeContext: Failed to initialize context for LUN #0! (interface 1343225856).
2020-11-25 09:33:05 -> usbHsFsDriveInitializeContext: Failed to initialize any LUN/filesystem contexts! (interface 1343225856).
2020-11-25 09:33:05 -> usbHsFsAddDriveContextToList: Freed drive context buffer.
2020-11-25 09:33:05 -> usbHsFsUpdateDriveContexts: Failed to add drive with ID 1343225856 to drive context list.
2020-11-25 09:33:05 -> usbHsFsUpdateDriveContexts: Added 0 drive context(s).
2020-11-25 09:33:05 -> usbHsFsDriveManagerThreadFunc: Interface state change event triggered.
2020-11-25 09:33:05 -> usbHsFsUpdateDriveContexts: Current drive count: 0.
2020-11-25 09:33:19 -> usbHsFsCloseDriveManagerThread: Signaling drive manager thread exit event...
2020-11-25 09:33:19 -> usbHsFsDriveManagerThreadFunc: Exit event triggered.
2020-11-25 09:33:19 -> usbHsFsCloseDriveManagerThread: Thread successfully closed.

**Comment by DarkMatterCore** [comment]
@duckbill007 Thanks a lot. Now I know what's happening. The BOT driver is sending a SCP INQUIRY command and expects 0x24 bytes in return, but your UMS device sends instead a CSW with a `Passed` status as if the data transfer already finished. This is called an "unexpected CSW" in BOT docs.

```
2020-11-25 09:33:05 -> usbHsFsScsiTransferCommand: Data from unexpected CSW (interface 1343225856, LUN 0):
55534253928b88340000000000
```

* `55 53 42 53` represents the CSW signature ("USBS").
* `92 8B 88 34` represents the CSW tag, which should match the one from the sent CBW. And it does, because otherwise that message wouldn't have been logged.
* `00 00 00 00` is the data residue. Under normal circumstances, it should always be zero.
* `00` is the status code. This is where it's messed up. I was actually expecting an error code here, but we got a `Passed` status code instead.

There's a good chance your device just doesn't support the SCP INQUIRY command. `fsp-usb` didn't rely on it, which explains how your drive worked with it. I'll add an exception for that, then upload a new build for you to test.

Then again, thanks a lot.

**Comment by duckbill007** [comment]
Yes. It is a quite old USB drive from Corsair. 8Gb in size, formatted as exfat for tests.

**Comment by DarkMatterCore** [comment]
@duckbill007 Alright, here's the new test build. Please let me know how it goes. :)

[libusbhsfs-example.zip](https://github.com/DarkMatterCore/libusbhsfs/files/5594985/libusbhsfs-example.zip)

**Comment by duckbill007** [comment]
Tried 3 times. All times all works ok.
https://dropmefiles.com/pMARU

**Comment by fennectech** [comment]
ill give it the ipod :shrek:

**Comment by DarkMatterCore** [comment]
@duckbill007 Thanks! :)

I just noticed the Inquiry command doesn't fail sometimes - there's an instance in which Inquiry data is properly sent by your flash drive. Makes me wonder if it's something related to delays. I'll see what I can do about that, then send you another build.

**Comment by duckbill007** [comment]
Any updates?

**Comment by DarkMatterCore** [comment]
@duckbill007 Yes. Please try this build - it's based on commit 75f53bf707053f6b070ec3c8fca53b9ef9ccece5.

[libusbhsfs-example.zip](https://github.com/DarkMatterCore/libusbhsfs/files/5606795/libusbhsfs-example.zip)

**Comment by DarkMatterCore** [comment]
A Request Sense command should now be issued after receiving a valid, but unexpected CSW.

**Comment by duckbill007** [comment]
Seems to work

**Comment by DarkMatterCore** [comment]
@duckbill007 Excellent, thanks! Could you please send me your latest logfile?

**Comment by duckbill007** [comment]
Here it is:
https://dropmefiles.com/EjWpB

**Comment by DarkMatterCore** [comment]
Fixed in a1cd27477d6c3bae75a24e57484a3e85dfedc988.
